File: Minimum Subarray.py
class Solution:
    """
    @param nums: A list of integers
    @return: An integer denote the sum of maximum subarray
    """
    def minSubArray(self, nums):
        # write your code here
        if not nums:
            return 0

        globalMin = nums[0]
        localMin = nums[0]

        for i in range(1, len(nums)):
            localMin = min(nums[i], localMin + nums[i])
            globalMin = min(localMin, globalMin)

        return globalMin


if __name__ == '__main__':
    a = Solution()
    nums = [-2, 2, -3, 4, -1, 2, 1, -5, 3]
    print a.minSubArray(nums)

File: java edition/minimumSubArray.java
import java.util.*;
public class minimumSubArray{
    public static void main(String[] args){
        ArrayList<Integer> test = new ArrayList<Integer>();
        test.add(1);
        test.add(-1);
        test.add(-2);
        test.add(1);
        System.out.println(minSubArray(test));
    }

    public static int minSubArray(ArrayList<Integer> nums) {
        int count = nums.size();
        if(nums.size() == 0 || nums == null){
            return 0;
        }

        ArrayList<Integer> newNums = new ArrayList<Integer>();
        for(Integer i : nums){
            newNums.add(-i);
        }

        int sum = 0;
        int max = Integer.MIN_VALUE;
        int minSum = 0;

        for(int i = 0; i < count; i++){
            System.out.println(newNums.get(i));
            sum += newNums.get(i);
            max = Math.max(max, sum - minSum);
            minSum = Math.min(minSum,sum);
        }
        return -max;
        // write your code
    }
}
